File: Recursion&Backtracking/permutation_II.py
'''Given a collection of numbers, nums, that might contain duplicates, return all possible unique permutations in any order.

 

Example 1:

Input: nums = [1,1,2]
Output:
[[1,1,2],
 [1,2,1],
 [2,1,1]]
Example 2:

Input: nums = [1,2,3]
Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]'''


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        used = [False]*len(nums)
        results = []
        path = []
        def backtrack(path):
            if len(path) == len(nums):
                results.append(path[:])
                return
            else:
                for i in range(len(nums)):
                    if used[i]:
                        continue
                    if i>0 and nums[i]==nums[i-1] and not used[i-1]:
                        continue
                    path.append(nums[i])
                    used[i] = True
                    backtrack(path)
                    path.pop()
                    used[i] = False
        backtrack(path)
        return results

File: Recursion&Backtracking/permutation_II.cpp
class Solution {
  public:
    void backtrack(string &s, vector<bool> used, string elements, vector<string> &result)
    {
        if(elements.size()==s.size())
        {
            result.push_back(elements);
            return;
        }
        for(int i=0; i<s.size(); i++)
        { 
            if (used[i])
                continue;
            if (i>0 && s[i-1] == s[i] && !used[i-1])
                continue;
            used[i]=true;
            elements.push_back(s[i]);
            backtrack(s, used, elements, result);
            elements.pop_back();
            used[i]=false;
        }
        return;
    }
    
    vector<string> findPermutation(string &s) {
        // Code here there
        vector<string> result;
        string elements;
        vector<bool> used(s.size(), false);
        sort(s.begin(), s.end());
        backtrack(s, used, elements, result);
        return result;
        
    }
};
